raiden/log_config.py

```diff
@@ -1,35 +1,113 @@
+import sys
+import logging
 import logging.config
-import structlog
-from typing import Dict
+from typing import Dict, Tuple, List
 
+import structlog
 
 DEFAULT_LOG_LEVEL = 'INFO'
+MAX_LOG_FILE_SIZE = 5 * 1024 * 1024
+
+
+def match_list(module_rule: Tuple[List[str], str], logger_module: str) -> Tuple[int, str]:
+    logger_modules_split = logger_module.split('.')
+
+    modules_split: List[str] = module_rule[0]
+    level: str = module_rule[1]
+
+    if logger_modules_split == modules_split:
+        return sys.maxsize, level
+    else:
+        num_modules = len(modules_split)
+        if logger_modules_split[:num_modules] == modules_split:
+            return num_modules, level
+        else:
+            return 0, None
+
+
+def get_log_level(
+    module_rules: List[Tuple[List[str], str]],
+    logger_module: str,
+    default_log_level: str = DEFAULT_LOG_LEVEL,
+) -> str:
+    best_match_length = 0
+    best_match_level = default_log_level
+
+    for module in module_rules:
+        match_length, level = match_list(module, logger_module)
+
+        if match_length > best_match_length:
+            best_match_length = match_length
+            best_match_level = level
+
+    return best_match_level
+
+
+class PerModuleFilteredRenderer(structlog.dev.ConsoleRenderer):
+    """ A Renderer that filters debug messages on a per-module basis. """
+    def __init__(self, log_level_config, colors=True):
+        super().__init__(colors=colors)
+        self._log_rules = [
+            (logger.split('.'), level)
+            for logger, level in log_level_config.items()
+        ]
+
+    def __call__(self, _, __, event_dict):
+        log_level_per_rule = get_log_level(self._log_rules, event_dict.get('logger', ''))
+        log_level_event = event_dict.get('level', DEFAULT_LOG_LEVEL).upper()
 
+        log_level_per_rule_numeric = getattr(logging, log_level_per_rule.upper(), 10)
+        log_level_event_numeric = getattr(logging, log_level_event.upper(), 10)
 
-def _get_log_handler(formatter, log_file):
+        # Drop the event when the log level is lower then the threshold
+        if log_level_event_numeric < log_level_per_rule_numeric:
+            # we cannot use `raise DropEvent` here, otherwise the event won't be logged
+            # in other renderers
+            return ''
+
+        return super().__call__(_, __, event_dict)
+
+
+def _get_log_handler(formatter: str, log_file: str, log_level: str) -> Dict:
```

Comment:
here and in the other function can you get more specific on the `Dict` return type. iow `Dict[str, Dict[str, Union[str, formatter_type]]]`

where formatter_type is whatever the type of `formatter` is. If it's also `str` then lose the `Union`.

Reply:
Yeah, but I'm not sure if this is reasonable at the moment. We don't have mypy checks, so it's mostly the arguments we want annotations for.